### PR TITLE
KAFKA-14170: Fix NPE in the deleteTopics() code path of KRaft Controller

### PR DESCRIPTION
**Pull request by akhileshchg**
KAFKA-14170: Fix NPE in the deleteTopics() code path of KRaft Controller

While replaying the `RemoveTopicRecord` in the Controller, we remove all the
in-memory states for the topic. While removing any offline partition data, if
there are any offline partitions in the system, we wrongly look up the offline
partition count for the topic we're removing. Fixed this lookup and added an
integration test.

While adding an integration test, found a bug in MetadataDelta#replay code
while processing `RemoveTopicRecord`. Fixed that as well.

### CODE REVIEW

File: core/src/test/scala/unit/kafka/server/DeleteTopicsRequestTest.scala [hunk 1]
```diff
@@ -32,10 +32,45 @@ import org.apache.kafka.common.requests.MetadataResponse
...
+    val offlineTopic = "topic-1"
+    val onlineTopic = "topic-2"
+    createTopicWithAssignment(offlineTopic, Map[Int, Seq[Int]](0 -> Seq(0)))
+    createTopicWithAssignment(onlineTopic, Map[Int, Seq[Int]](0 -> Seq(1)))
+    killBroker(0)
+    ensureConsistentKRaftMetadata()
+
+    // Ensure one topic partition is offline.
+    TestUtils.waitUntilTrue(() => {
+      aliveBrokers.head.metadataCache.getPartitionInfo(onlineTopic, 0).exists(_.leader() == 1)
```
Comment (dengziming) [inline core/src/test/scala/unit/kafka/server/DeleteTopicsRequestTest.scala]: We miss a "&&" between these 2 expressions.

File: core/src/test/scala/unit/kafka/server/DeleteTopicsRequestTest.scala [hunk 2]
```diff
@@ -32,10 +32,45 @@ import org.apache.kafka.common.requests.MetadataResponse
...
+      aliveBrokers.head.metadataCache.getPartitionInfo(offlineTopic, 0).exists(_.leader() == -1)
```
Comment (dengziming) [inline core/src/test/scala/unit/kafka/server/DeleteTopicsRequestTest.scala]: nit: using MetadataResponse.NO_LEADER_ID rather than -1

File: metadata/src/main/java/org/apache/kafka/controller/BrokersToIsrs.java [hunk 3]
```diff
@@ -191,7 +191,7 @@ void update(Uuid topicId, int partitionId, int[] prevIsr, int[] nextIsr,
     void removeTopicEntryForBroker(Uuid topicId, int brokerId) {
         Map<Uuid, int[]> topicMap = isrMembers.get(brokerId);
         if (topicMap != null) {
-            if (brokerId == NO_LEADER) {
+            if (brokerId == NO_LEADER && topicMap.containsKey(topicId)) {
                 offlinePartitionCount.set(offlinePartitionCount.get() - topicMap.get(topicId).length);
```
Comment (dengziming) [inline metadata/src/main/java/org/apache/kafka/controller/BrokersToIsrs.java:195]: Does this mean the `offlinePartitionCount` is no longer accurate? we don't update `offlinePartitionCount` when we delete a topic.

File: core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala [hunk 4]
```diff
@@ -173,11 +173,11 @@ abstract class KafkaServerTestHarness extends QuorumTestHarness {
     adminClientConfig: Properties = new Properties
   ): scala.collection.immutable.Map[Int, Int] = {
     if (isKRaftTest()) {
-      resource(createAdminClient(brokers, listenerName, adminClientConfig)) { admin =>
+      resource(createAdminClient(aliveBrokers, listenerName, adminClientConfig)) { admin =>
```
Comment (dengziming) [inline core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala]: Is this change necessary? I think AdminClient should be robust enough to recognize dead and alive brokers and handle it properly.

**Review by dengziming — COMMENTED** [review]
Good catch @akhileshchg , I left some comments.

File: metadata/src/main/java/org/apache/kafka/controller/BrokersToIsrs.java (hunk 3 above)
Comment (cmccabe) [reply metadata/src/main/java/org/apache/kafka/controller/BrokersToIsrs.java:195]: Hmm... We do update `offlinePartitionCount` when we delete a topic, right? Check `ReplicationControlManager#replay(RemoveTopicRecord record)`

The change here is to avoid a null pointer exception if `topicMap.get(topicId) == null`, which I believe is a correct fix.

File: core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala (hunk 4 above)
Comment (cmccabe) [reply core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala]: I agree. Let's revert the changes in this file.

Comment (akhileshchg) [reply core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala]: The problem is we need at least one change in `deleteTopic()`. Because it checks to see if metadata is propagated to all the brokers in `(brokers, topic, 0)`. This will fail if one of the `brokers` passed to `waitForAllPartitionsMetadata` is dead.

File: metadata/src/main/java/org/apache/kafka/controller/BrokersToIsrs.java (hunk 3 above)
Comment (akhileshchg) [reply metadata/src/main/java/org/apache/kafka/controller/BrokersToIsrs.java:195]: That's right, @cmccabe. We still correctly update the offline partition count.

File: core/src/test/scala/unit/kafka/server/DeleteTopicsRequestTest.scala (hunk 1 above)
Comment (akhileshchg) [reply core/src/test/scala/unit/kafka/server/DeleteTopicsRequestTest.scala]: Thanks for the catch.

File: core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala (hunk 4 above)
Comment (akhileshchg) [reply core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala]: Let me know your thoughts. I removed the check while giving it to create adminClient but still kept in the places we check for metadata propagation.

Comment (cmccabe) [reply core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala]: If that is the case, then you should wait until all brokers are unfenced before proceeding in the test

**Comment by akhileshchg** [comment]
@dengziming @cmccabe I addressed your comments. I have question regarding KafkaTestHarness changes. Please let me know your thoughts.

File: core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala (hunk 4 above)
Comment (cmccabe) [reply core/src/test/scala/unit/kafka/integration/KafkaServerTestHarness.scala]: I looked at this again, and I think using aliveBrokers is correct.

**Review by cmccabe — APPROVED** [review]
LGTM
